exonum/src/messages2/mod.rs

```diff
@@ -0,0 +1,282 @@
+pub struct Message<T> {
+    //TODO: inner T duplicate data in SignedMessage, we can use owning_ref,
+    //if our serialization format allows us (ECR-2315).
+    payload: T,
+    #[serde(with = "HexStringRepresentation")]
+    message: SignedMessage,
+}
+
+impl<T: ProtocolMessage> Message<T> {
+    /// Creates new instance of the message.
+    pub(in messages) fn new(payload: T, message: SignedMessage) -> Message<T> {
+        Message { payload, message }
+    }
+
+    /// Returns hash of the full message.
+    pub fn hash(&self) -> Hash {
+        hash(self.message.raw())
+    }
+
+    /// Returns a serialized buffer.
+    pub fn serialize(self) -> Vec<u8> {
+        self.message.raw
+    }
+
+    /// Return reference to payload.
+    pub fn payload(&self) -> &T {
+        &self.payload
+    }
+
+    /// Return reference to signed message.
+    pub(crate) fn signed_message(&self) -> &SignedMessage {
+        &self.message
+    }
+
+    /// Returns public key of the message creator.
+    pub fn author(&self) -> PublicKey {
+        self.message.author()
+    }
+}
+
+impl fmt::Debug for ServiceTransaction {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.debug_struct("Transaction")
+            .field("message_id", &self.transaction_id)
+            .field("payload_len", &self.payload.len())
+            .finish()
+    }
+}
+
+impl<T> ToHex for Message<T> {
+    fn write_hex<W: ::std::fmt::Write>(&self, w: &mut W) -> ::std::fmt::Result {
+        self.message.raw().write_hex(w)
+    }
+
+    fn write_hex_upper<W: ::std::fmt::Write>(&self, w: &mut W) -> ::std::fmt::Result {
+        self.message.raw().write_hex_upper(w)
+    }
+}
+
+impl<X: ProtocolMessage> FromHex for Message<X> {
+    type Error = Error;
+
+    fn from_hex<T: AsRef<[u8]>>(v: T) -> Result<Self, Error> {
+        let bytes = Vec::<u8>::from_hex(v)?;
+        let protocol = Protocol::deserialize(SignedMessage::from_raw_buffer(bytes)?)?;
+        ProtocolMessage::try_from(protocol)
+            .map_err(|_| format_err!("Couldn't deserialize message."))
+    }
+}
+
+impl<T: ProtocolMessage> AsRef<SignedMessage> for Message<T> {
+    fn as_ref(&self) -> &SignedMessage {
+        &self.message
+    }
+}
+
+impl<T: ProtocolMessage> AsRef<T> for Message<T> {
+    fn as_ref(&self) -> &T {
+        &self.payload
+    }
+}
+
+impl<T> From<Message<T>> for SignedMessage {
+    fn from(message: Message<T>) -> Self {
+        message.message
+    }
+}
+
+impl<T: ProtocolMessage> Deref for Message<T> {
+    type Target = T;
+    fn deref(&self) -> &T {
+        &self.payload
+    }
+}
+
+impl<T: ProtocolMessage> StorageValue for Message<T> {
+    fn into_bytes(self) -> Vec<u8> {
+        self.message.raw
+    }
+
+    fn from_bytes(value: Cow<[u8]>) -> Self {
+        let message = SignedMessage::from_vec_unchecked(value.into_owned());
+        //TODO: Remove additional deserialization [ECR-2315]
```

Comment:
Please add a dot at the end of the comment.

Additionally I would love to see space between "//` and "TODO".